### ISSUE TITLE
PermissionChecker does not handle list of permission correctly

### ISSUE THREAD

**Issue by RomekCDPR** [quarkusio/quarkus#45224]
### Describe the bug

When method is annotated with `@PermissionsAllowed` with multiple values `PermissionChecker` is not correctly matched.

```java
@PermissionsAllowed(value = {"read", "read:all"})
@Override
public CompletionStage<Response> getUserPlatformProducts(String userId) {
    // ...
}

@PermissionChecker("read")
boolean canRead(SecurityIdentity identity, String userId) {
    return identity.getPrincipal().getName().equals(userId);
}
```

### Expected behavior

Checker should be called when request has `write` permission.

### Actual behavior

```
java.lang.RuntimeException: java.lang.RuntimeException: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
	[error]: Build step io.quarkus.security.deployment.SecurityProcessor#gatherSecurityChecks threw an exception: java.lang.RuntimeException: Found @PermissionChecker annotation instance that authorize the 'read' permission, however
no @PermissionsAllowed annotation instance requires this permission
```

**Comment by sberyozkin** [comment]
@RomekCDPR Can you please update the example to show where the `write` permission is listed ?

**Comment by RomekCDPR** [comment]
I've removed `write` permission and updated exception for clarity.

**Comment by michalvavrik** [comment]
@RomekCDPR good point, I need to tweak validation bit, thanks for reporting this. will have a look in next 2 days

**Comment by michalvavrik** [comment]
For anyone without context, this is validation inside a security processor and build fails.

**Comment by michalvavrik** [comment]
> When method is annotated with @PermissionsAllowed with multiple values PermissionChecker is not correctly matched.

BTW we have plenty of tests, maybe this may need to rephrase. I'll try to map what are exact situations when this validation fails when I create reproduce.

**Comment by michalvavrik** [comment]
@RomekCDPR I checked and I can reproduce the issue. In the `@PermissionChecker` javadoc you can see that we didn't explicitly define how will permission actions match between the checker annotation and `@PermissionsAllowed` annotation https://github.com/quarkusio/quarkus-security/blob/main/src/main/java/io/quarkus/security/PermissionChecker.java. What I did was to make exact match: `@PermissionsAllowed("read:action")` will only match `@PermissionChecker("read:action")`. I thought it was the most straightforward way.

Now, there is nothing wrong about what you did `@PermissionsAllowed(value = {"read", "read:all"})`, according to the docs https://github.com/quarkusio/quarkus-security/blob/main/src/main/java/io/quarkus/security/PermissionsAllowed.java#L143 one of `read` and `read:all` is needed.

So it doesn't make much of a sense because if you have `read`, you will also have `read:all` and any other action (like `read:me`), but it makes sense for the `@PermissionChecker` I guess, because you can still have checker granting `read` and not `@PermissionChecker("read:all")`.

Hope you are not lost yet :-), I'll fix the validation.

**Comment by michalvavrik** [comment]
Honestly, I am still bit on a fence whether `@PermissionsAllowed(value = {"read", "read:all"})` makes sense. If you say: either have `read` permission, or `read` permission with action `all`, then `all` feels like surplus to requirement and probably signals someone misunderstood docs. I'll think about it, maybe improved validation exception message would do as well.

**Comment by sberyozkin** [comment]
It should be a direct string match, if it is `read:something` that it can only match `read:something`, which is what you did and it is fine.

**Comment by michalvavrik** [comment]
I think you are right @sberyozkin , in fact, when I added a validation exception, I found out that we have plenty of tests like these, for example:

https://github.com/quarkusio/quarkus/blob/0ea2a40e56fb17c9dba65179302fd398cd125a7f/extensions/security/deployment/src/test/java/io/quarkus/security/test/permissionsallowed/MethodLevelCustomPermissionsAllowedTest.java#L282

Only difference is that all specify first `"three:c", "three"`, if only I switched order somewhere, I could had discover this validation bug sooner.

**Comment by sberyozkin** [comment]
Even though it appears that `read:all` is a duplicate of a more powerful `read`, we can't really make such conclusions during the match, in some systems `read:all` means a complete admin read level access while `read` - some limited access. Or these permissions may not even intersect :-), so the safest way for us is just to do the precise match.

May be, if proves necessary, we can support something like `@PermissionChecker("read:*")` (wildcard for actions only), with this checker called to check `read:a`, `read:b`, etc, but not `read`

**Comment by michalvavrik** [comment]
I have been thinking about this since I wrote here first comment @sberyozkin , and I think supporting actions for `@PermissionChecker` methods will require more discussion. What I didn't remember is that there is already validation in place that breaks the build if you try to use permission actions in `@PermissionChecker` value:

- https://github.com/quarkusio/quarkus/blob/2ac009a43354870e96ebb3fa7d0b0563fc3afd08/extensions/security/deployment/src/main/java/io/quarkus/security/deployment/PermissionSecurityChecks.java#L147

What I am worried about is intersection between permission actions checked by a permission checker and the same permission actions when it is not checked by a permission checker and inclusive flag is set to true. I'd like to be careful and fix just this validation bug, we can open a feature request for supporting permission checker actions. I think it will need a discussion what is expected behavior and it won't be trivial. Hope it's alright.

**Comment by michalvavrik** [comment]
I will summarize here my takeaways from https://github.com/quarkusio/quarkus/pull/45256:

- ATM build fails when actions are used in `@PermissionChecker` method, that is well documented behavior
- scenario with `@PermissionsAllowed(value = {"read", "read_all"})` is one case I could fine that provides confusing build-time validation exception
- plan is to support (so called) "actions" in `@PermissionChecker` based on the string equality
- Quarkus Security API must be updated before that happens

Note: this is basically what I was doing yesterday, so I already know changes won't be trivial. Right now `@PermissionAllowed({"read", "read:all", "read:some"})` is mapped to one permission instance (docs: https://github.com/quarkusio/quarkus-security/blob/8eb9e15216408419e580049c3a4f034a3c980fb4/src/main/java/io/quarkus/security/PermissionsAllowed.java#L143). However for a permission checker, 3 instances are going to be required. Once docs is easy to understand to users, I'll attempt to implemented this, but this is not the highest on my list. No user required it so far (and I believe neither this issue requires it at all), so I'll prioritize other issues. Thanks

**Comment by sberyozkin** [comment]
@michalvavrik Hi Michal

>plan is to support (so called) "actions" in @PermissionChecker based on the string equality

IMHO we should not be talking about actions at all when dealing with `PermissionAllowed` and `PermissionChecker`, this is only going to be confusing.  I've tried to summarize why in #45256. Let users who deal with custom Java Permissions map `read:all` into a permission name, permission actions because this is what custom Java Permission requires. 

IMHO there is no need to get the `PermissionChecker` complicated beyond the String comparison.

**Comment by sberyozkin** [comment]
>Right now @PermissionAllowed({"read", "read:all", "read:some"}) is mapped to one permission instance (docs: https://github.com/quarkusio/quarkus-security/blob/8eb9e15216408419e580049c3a4f034a3c980fb4/src/main/java/io/quarkus/security/PermissionsAllowed.java#L143). However for a permission checker, 3 instances are going to be required.

There is no `Permission` instance for users dealing `PermissionChecker` - it is an alternative, `QuarkusPermission` which supports checkers is an implementation detail. IMHO it is important that we keep `Java Permission` out of the user's picture in this case.

The other thing is that one checker is also required in this case by default. 3 checkers are only required if `PermissionAllowed` has an `inclusive` flag activated.

Thanks

**Comment by michalvavrik** [comment]
sure @sberyozkin , I have thought about it in the meanwhile and I think most of my worries can be addressed by a simple rule that when `@PermissionAllowed({"read", "read:all", "read:some"}) ` has at least one permission checker, let say for `@PermissionChecker("read:all")`, then there **must** exist checker for `read` and `read:some`. My issue was that we shouldn't allow to grant `read:some` and `read` by a `SecurityIdentity` permission because it will make things complicated (I tried to explain why before, I can retry if not clear).

**Comment by michalvavrik** [comment]
So, yes, I agree with the string comparison.

**Comment by sberyozkin** [comment]
Hi Michal @michalvavrik 

>simple rule that when @PermissionAllowed({"read", "read:all", "read:some", "write"}) has at least one permission checker, let say for @PermissionChecker("read:all"), then there must exist checker for read and read:some

Why exactly ? 

By default, `@PermissionAllowed({"read", "read:all", "read:some", "write"})` is non inclusive, i.e, only one of `read`, `read:all`, `read:some` or `write` must be satisfied. We don't know the semantics of these 3 read variations. `@PermissionChecker("read:all")` should be enough, as long as it returns `true`.

>My issue was that we shouldn't allow to grant read:some and read by a SecurityIdentity permission because it will make things complicated

Do you mean that if only `@PermissionChecker("read:all")` is available, we should not grant access to the request which only have `read` or `read:some` permissions ? It should be indeed the case if it is inclusive, but otherwise  `@PermissionChecker("read:all")` should be enough, as long as it returns `true`.

Thanks

**Comment by michalvavrik** [comment]
> Hi Michal [@michalvavrik](https://github.com/michalvavrik)
> 
> > simple rule that when @PermissionAllowed({"read", "read:all", "read:some", "write"}) has at least one permission checker, let say for [@permissionchecker](https://github.com/permissionchecker)("read:all"), then there must exist checker for read and read:some
> 
> Why exactly ?
> 
> By default, `@PermissionAllowed({"read", "read:all", "read:some", "write"})` is non inclusive, i.e, only one of `read`, `read:all`, `read:some` or `write` must be satisfied. We don't know the semantics of these 3 read variations. `@PermissionChecker("read:all")` should be enough, as long as it returns `true`.

Yes, by default this annotation instance won't be inclusive, but if you add `inclusive=true` than it becomes complicated. I thought the requirement on what checks `read`, `read:all` and `read:some` shouldn't be dependent on the `inclusive` flag. That is why. I can live with that though, if you agree that it will be required at least for `inclusive=true`.

> > My issue was that we shouldn't allow to grant read:some and read by a SecurityIdentity permission because it will make things complicated
> 
> Do you mean that if only `@PermissionChecker("read:all")` is available, we should not grant access to the request which only have `read` or `read:some` permissions ?

No, I say if the `@PermissionChecker("read:all")` is not available for the `@PermissionAllowed(value = {"read", "read:all", "read:some", "write"}, inclusive = true)`, but `@PermissionChecker("read:some")` is detected, the build should fail and validation exception should be thrown.

>  It should be indeed the case if it is inclusive, but otherwise `@PermissionChecker("read:all")` should be enough, as long as it returns `true`.

Ok, I agree on that.

**Comment by sberyozkin** [comment]
@michalvavrik Sure, we are in complete agreement. `inclusive` flag determines if every listed allowed permission must have a matching checker or not. If `inclusive=true` then if there are 10 permissions listed then should be be 10 matching checkers.

May be, for the future, we can allow some optimizations like

`@PermissionAllowed({"a:1", "a:2"})`

and 

`@PermissionChecker("a:*")` 

but this option would require a dedicated parameter for the checker to know which permission is actually being checked.
